controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -18,30 +20,47 @@
 import io.pravega.controller.store.stream.tables.Data;
 import io.pravega.controller.store.stream.tables.HistoryRecord;
 import io.pravega.controller.store.stream.tables.IndexRecord;
-import io.pravega.controller.store.stream.tables.SegmentRecord;
 import io.pravega.controller.store.stream.tables.State;
 import io.pravega.controller.store.stream.tables.TableHelper;
-import io.pravega.client.stream.StreamConfiguration;
 import lombok.extern.slf4j.Slf4j;
 import lombok.val;
+import org.apache.commons.lang.SerializationUtils;
 import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
 
 import java.util.AbstractMap;
 import java.util.AbstractMap.SimpleImmutableEntry;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 @Slf4j
 public abstract class PersistentStreamBase<T> implements Stream {
 
+    private static final Map<State, List<State>> STATE_TRANSITIONS;
```

Comment:
enum StateTransitions {
    UNKNOWN(State.UNKNOWN, State.CREATING),
    CREATING(State.CREATING, State.ACTIVE),
    ACTIVE(State.ACTIVE, State.SCALING, State.SEALING, State.SEALED, State.UPDATING),
    SCALING(State.SCALING, State.ACTIVE),
    UPDATING(State.UPDATING, State.ACTIVE),
    SEALING(State.SCALING, State.SEALED),
    SEALED(State.SEALED);
    
    State[] transitions;
    
    StateTransitions(State... states) {
        this.transitions = states;
    }
}